### PR TITLE
Compatibility

### PR DESCRIPTION
**Pull request by chaoticlonghair**
- Compatibility: discard the verified transactions saved in wal by previous versions.
- Compatibility: process the proposals saved in wal by previous versions.
- Take all txs from BlockTxs in case of the proposals include them can not verified forever.
- If auth doesn't respond for a long while, bft do not wait it.

### CODE REVIEW

File: src/core/cita_bft.rs [hunk 1]
```diff
@@ -1112,12 +1114,28 @@ impl Bft {
...
+        let mut hash_v0_20_x = None;
+        let mut csp_msg = if let Ok(csp_msg) = Message::try_from(body) {
             csp_msg
         } else {
             return Err(EngineError::UnexpectedMessage);
         };
-        let result = csp_msg.clone().take_compact_signed_proposal();
+        let result = csp_msg.clone().take_compact_signed_proposal().or_else(|| {
+            // Only process the old version proposals from wal
+            if !wal_flag {
```
Comment (u2) [inline src/core/cita_bft.rs:1149]: 这里只要有wal就一定是原来的旧版本的 proposal？

Comment (chaoticlonghair) [reply src/core/cita_bft.rs:1149]: No!

Pseudo code:

```rust
let new_result = parse_as_new_proposal(data);
let result = if new_result.is_err() {
    if from_wal {
        let old_result = parse_as_old_proposal(data);
        old_result.unwrap()
    } else {
        throw_exception();
    }
} else {
    new_result.unwrap()
};
```

Comment (u2) [reply src/core/cita_bft.rs:1149]: Get it.

File: src/core/wal.rs [hunk 2]
```diff
@@ -23,15 +23,17 @@ use std::str;
 
 const DELETE_FILE_INTERVAL: usize = 3;
 
+#[derive(Debug, Clone, Copy)]
 pub enum LogType {
+    Skip = !0,
```
Comment (u2) [inline src/core/wal.rs:28]: `!0` is `not 0`? What does it stand for?

File: src/core/cita_bft.rs [hunk 3]
```diff
@@ -1289,15 +1283,18 @@ impl Bft {
...
                     "new_proposal BLOCKTXS get height {}, self height {}",
-                    *height,
+                    height,
                     self.height
                 );
-                if *height == self.height - 1 {
+                if height == self.height - 1 {
                     flag = true;
-                    block.set_body(blocktxs.get_body().clone());
+                    // Take all transactions in case of the proposal include them can't be verified
```
Comment (u2) [inline src/core/cita_bft.rs]: Need to add the explanation that it must be `take` if compact-block-relay. If it's a full-block-relay proposal either `take` or `get` is ok.

Comment (u2) [reply src/core/cita_bft.rs]: There are some grammar errors in the first sentence.

File: src/core/cita_bft.rs [hunk 4]
```diff
@@ -62,19 +62,43 @@ pub enum BftTurn {
     Timeout(TimeoutInfo),
 }
 
-#[derive(Debug, PartialEq, PartialOrd, Clone, Copy)]
+#[derive(Debug, Clone, Copy)]
 enum VerifiedBlockStatus {
-    Ok = 1,
-    Failed = -1,
-    Undo = 0,
+    Ok,
```
Comment (u2) [inline src/core/cita_bft.rs:67]: OK?

Comment (chaoticlonghair) [reply src/core/cita_bft.rs:67]: Refer to [std::result::Result]:
```rust
pub enum Result<T, E> {
    Ok(T),
    Err(E),
}
```

[std::result::Result]: https://doc.rust-lang.org/std/result/enum.Result.html

File: src/core/wal.rs (hunk 2 above)
Comment (chaoticlonghair) [reply src/core/wal.rs:28]: Set all bits to `1` without specify the actual type (more flexible).

File: src/core/cita_bft.rs (hunk 3 above)
Comment (chaoticlonghair) [reply src/core/cita_bft.rs]: IMHO, if it is a full-block-relay proposal, there must be `take`, too.

File: src/core/cita_bft.rs (hunk 4 above)
Comment (u2) [reply src/core/cita_bft.rs:67]: ok

File: src/core/cita_bft.rs (hunk 3 above)
Comment (chaoticlonghair) [reply src/core/cita_bft.rs]: I will change this sentence, but I do not think it has grammar errors.

**Review by u2 — APPROVED** [review]

**Comment by u2** [comment]
@yangby-cryptape @rink1969 Any more discussion? @jerry-yu

**Review by jerry-yu — APPROVED** [review]

**Pull request merged**